dvc/render/data.py

```diff
@@ -0,0 +1,199 @@
+from typing import Callable, Dict, List, Set, Union
+
+from funcy import first, reraise
+
+from dvc.exceptions import DvcException
+from dvc.render.base import INDEX_FIELD, REVISION_FIELD
+
+
+class FieldsNotFoundError(DvcException):
+    def __init__(self, expected_fields, found_fields):
+        self.expected_fields = expected_fields
+        self.found_fields = found_fields
+
+        expected_str = ", ".join(expected_fields)
+        found_str = ", ".join(found_fields)
+        super().__init__(
+            f"Could not find all provided fields ('{expected_str}') "
+            f"in data fields ('{found_str}')."
+        )
+
+
+class PlotDataStructureError(DvcException):
+    def __init__(self):
+        super().__init__(
+            "Plot data extraction failed. Please see "
+            "https://man.dvc.org/plots for supported data formats."
+        )
+
+
+def _filter_fields(fields=None) -> Callable:
+    def process(datapoints: List[Dict]) -> List[Dict]:
+
+        if not fields:
+            return datapoints
+        assert isinstance(fields, set)
+
+        new_data = []
+        for data_point in datapoints:
+            new_dp = data_point.copy()
+
+            keys = set(data_point.keys())
+            if keys & fields != fields:
+                raise FieldsNotFoundError(fields, keys)
+
+            to_del = keys - fields
+            for key in to_del:
+                del new_dp[key]
+            new_data.append(new_dp)
+        return new_data
+
+    return process
```

Comment:
As an alternative can keep flat functions and use `partial()` when making steps. Can also make them named with something like:
```python
def named_partial(name, func, *args, **kwargs):
    p = partial(func, *args, **kwargs)
    p.name = name
    return p

c.steps = [
    named_partial("find_data", _find_data, fields=self.props.get("fields"),
    ...
]
```
Or put names outside of callable objects:
```python
c.steps = [
    ("find_data", partial(_find_data, fields=self.props.get("fields")),
    ...
]

```

Reply:
Great idea. It would also make skip functionality very easy.

Reply:
Implemented names outside of partial, to keep mypy happy.